### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by xithan**

### CODE REVIEW

File: lib/smart_name/manipulate.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      old_name = old.to_name
+      new_name = new.to_name
+      return self if old_name.length > length
+      return replace_part(old_name, new_name) if old_name.simple?
+      return self unless include? old_name
+      replace_all_subsequences(old_name.parts, new_name.parts).to_name
+    end
+
+    def replace_part oldpart, newpart
+      ensure_simpliness oldpart, "Use 'replace' to replace junctions"
```
Comment (xithan) [inline lib/smart_name/manipulate.rb]: `replace` acts as the old `replace_part` and `replace_part` now replaces only parts

File: lib/smart_name/manipulate.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    # replace a subname
+    # keys are used for comparison
+    def replace old, new
+      old_name = old.to_name
+      new_name = new.to_name
+      return self if old_name.length > length
+      return replace_part(old_name, new_name) if old_name.simple?
+      return self unless include? old_name
+      replace_all_subsequences(old_name, new_name).to_name
+    end
```
Comment (ethn) [inline lib/smart_name/manipulate.rb]: I think you know I was joking when I said that anyone who moves code has to refactor it, right?

But this _rocks_!

**Review by ethn — APPROVED** [review]

File: lib/smart_name/manipulate.rb (hunk 2 above)
Comment (GerryG) [reply lib/smart_name/manipulate.rb:12]: It took me a while to find the old version, which is of course why you're not supposed to move and refactor, at least not in the same commit. This way makes way more sense, I wonder when the old logic was basically formed, probably goes pretty far back.
Are we sure we didn't change any behavior?

File: lib/smart_name/manipulate.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    def ensure_simpliness part, msg=nil
+      return if part.to_name.simple?
+      raise StandardError, "'#{part}' has to be simple. #{msg}"
+    end
+  end
+
+  # ~~~~~~~~~~~~~~~~~~~~ MISC ~~~~~~~~~~~~~~~~~~~~
+
+  # HACK. This doesn't belong here.
+  # shouldn't it use inclusions???
```
Comment (GerryG) [inline lib/smart_name/manipulate.rb:64]: Where is this used? I was thinking it was part of the context names stuff, but it isn't. Couldn't find any refs in smartname or wagn, is this used in wikirate maybe?

**Pull request merged**